src/sidebar/build-thread.js

```diff
@@ -305,86 +297,99 @@ const defaultOpts = {
 
   let thread = threadAnnotations(annotations);
 
-  // Mark annotations as visible or hidden depending on whether
-  // they are being edited and whether they match the current filter
-  // criteria
-  const shouldShowThread = function (annotation) {
-    if (opts.forceVisible && opts.forceVisible.indexOf(id(annotation)) !== -1) {
-      return true;
-    }
-    if (opts.filterFn && !opts.filterFn(annotation)) {
-      return false;
-    }
-    return true;
-  };
+  if (hasSelection) {
+    // Remove threads (annotations) that are not selected
+    thread.children = thread.children.filter(
+      child => opts.selected.indexOf(child.id) !== -1
+    );
+  }
 
-  // When there is a selection, only include top-level threads (annotations)
-  // that are selected
-  if (opts.selected.length > 0) {
-    thread = Object.assign({}, thread, {
-      children: thread.children.filter(function (child) {
-        return opts.selected.indexOf(child.id) !== -1;
-      }),
-    });
+  if (threadsFiltered) {
+    // Remove threads not matching thread-level filters
+    thread.children = thread.children.filter(opts.threadFilterFn);
   }
 
-  // Set the visibility and highlight states of threads
-  thread = mapThread(thread, function (thread) {
-    let highlightState;
-    if (opts.highlighted.length > 0) {
-      const isHighlighted =
-        thread.annotation && opts.highlighted.indexOf(thread.id) !== -1;
-      highlightState = isHighlighted ? 'highlight' : 'dim';
+  // Set visibility for threads
+  thread = mapThread(thread, thread => {
+    let threadIsVisible = thread.visible;
+
+    if (!thread.annotation) {
+      threadIsVisible = false; // Nothing to show
+    } else if (annotationsFiltered) {
+      if (
+        hasForcedVisible &&
+        opts.forceVisible.indexOf(annotationId(thread.annotation)) !== -1
+      ) {
+        // This annotation may or may not match the filter, but we should
+        // make sure it is visible because it has been forced visible by user
+        threadIsVisible = true;
+      } else {
+        // Otherwise, visibility depends on whether it matches the filter
+        threadIsVisible = !!opts.filterFn(thread.annotation);
+      }
     }
-
-    return Object.assign({}, thread, {
-      highlightState: highlightState,
-      visible:
-        thread.visible &&
-        thread.annotation &&
-        shouldShowThread(thread.annotation),
-    });
+    return { ...thread, ...{ visible: threadIsVisible } };
   });
 
-  // Expand any threads which:
-  // 1) Have been explicitly expanded OR
-  // 2) Have children matching the filter
-  thread = mapThread(thread, function (thread) {
-    const id = thread.id;
-
-    // If the thread was explicitly expanded or collapsed, respect that option
-    if (opts.expanded.hasOwnProperty(id)) {
-      return Object.assign({}, thread, { collapsed: !opts.expanded[id] });
+  // Remove top-level threads which contain no visible annotations
```

Comment:
I've tried to hoist actions that _remove_ annotations/threads higher in the function but obviously this one depends on derived visibility...